### PR TITLE
sdr-dsp: NOAA APT decoder DSP pipeline (closes #477)

### LINKED ISSUE #477: APT decoder: FM demod → 11025 Hz resample → 2400 Hz AM envelope → sync detect
## Scope

Core signal processing for NOAA APT. Given FM-demodulated audio (from the existing FM demod path), produce a stream of synchronized scan lines at the APT line rate.

## Pipeline

\`\`\`
FM-demodulated audio (from sdr-radio, 48 kHz typical)
    │
    ▼
resample to 11025 Hz (polyphase, integer-samples-per-pixel friendly)
    │
    ▼
AM envelope demod of 2400 Hz subcarrier
    │  (Hilbert transform, or rectify + low-pass, or complex-mix then magnitude)
    ▼
sync detection (Sync A = 1040 Hz burst, Sync B = 832 Hz burst)
    │
    ▼
scan line output: 2080-pixel Vec<u8> stream, one per 500 ms
\`\`\`

## Implementation notes

- Lives in `sdr-dsp` (pure math, no I/O) or a new `sdr-apt` crate if it gets big
- Input rate is 48 kHz by convention; output samples per pixel is integer after resample to 11.025 or 22.050 kHz
- Sync detection is correlation-based — cross-correlate against known Sync A / Sync B patterns
- Handle the line-rate drift from Doppler (~ ±20 ppm over a pass) by searching a ±1-sample window each line
- Pre-recorded WAV samples from https://github.com/martinber/noaa-apt/tree/main/tests/ make good unit-test fixtures

## Deliverables

- `sdr_dsp::apt::AptDecoder` (or `sdr_apt::Decoder`) with `process(audio: &[f32]) -> Vec<AptLine>`
- `AptLine { samples: Vec<u8>, channel_a_sync_quality: f32, channel_b_sync_quality: f32, line_timestamp: ... }`
- Unit tests using known-good recorded passes — assert that N lines are extracted with sync quality above threshold

## Out of scope
- Scan-line assembly into a 2D image (next ticket)
- Telemetry wedge decode (later ticket)
- Doppler correction

Part of epic #468.

### PR DESCRIPTION
**Pull request by jasonherald**
## Summary

First sub-ticket of the NOAA APT epic (#468 / #477). Adds a pure `sdr-dsp::apt` module — no UI, no threading, no wiring into the signal path yet.

- `EnvelopeDetector` — rectify + LPF to pull the picture envelope off the 2400 Hz AM subcarrier
- `RealResampler` — thin real-audio wrapper over the existing complex `RationalResampler` so 48 kHz FM-demod audio can land on the internal 20 800 Hz grid
- `SyncDetector` — ±1 square-wave templates for Sync A (140 samples, 1040 Hz × 7) and Sync B (175 samples, 832 Hz × 7), matched with DC-invariant normalized cross-correlation
- `AptDecoder` — end-to-end stateful pipeline that chains the above into `AptLine { pixels: [u8; 2080], sync_quality, sync_channel, input_sample_index }`

## Design choices worth calling out

**Internal rate 20 800 Hz** — smallest rate that gives integer samples per pixel (5), integer cycles per sync tone (20 / 25), and keeps `2·f_subcarrier` (4800 Hz) below Nyquist so the envelope LPF cleanly removes the rectification harmonic. All downstream indexing is exact integer arithmetic — no fractional alignment headaches.

**Rectify + LPF instead of Hilbert magnitude** — traditional `noaa-apt` / `wxtoimg` approach, half the FIR work, adequate quality.

**Per-line min/max pixel normalization** — placeholder that always produces a visible image. Proper wedge-bar / telemetry-based calibration is a follow-up (tracked in the epic).

**Caller owns quality filtering** — every `AptLine` carries `sync_quality ∈ [0, 1]`; the decoder never swallows low-confidence lines on its own. UI code decides what to paint.

## Test plan

- [x] 27 new unit tests in `sdr_dsp::apt::tests` covering: envelope DC recovery from synthetic 2400 Hz AM, ramp envelope tracking, resampler tone fidelity + cross-chunk continuity, sync template shape / DC-invariance / noise rejection, chunked-vs-one-shot decoder equivalence, accumulator bounding under noise, end-to-end multi-line gradient preservation
- [x] `cargo test -p sdr-dsp` — all 259 tests pass (232 existing + 27 new)
- [x] `cargo clippy --all-targets --workspace -- -D warnings` — clean
- [x] `cargo fmt --all -- --check` — clean
- [x] `cargo build --workspace` — full workspace compiles

## What's not in this PR (follow-ups in the epic)

- [ ] #478 — wire the decoder into `sdr-radio` as a new module
- [ ] #479 — scrolling waterfall display of decoded lines in the UI
- [ ] #480 — telemetry-bar-based pixel calibration (replaces the per-line min/max normalization)
- [ ] #481-#484 — Sync B channel emission, false-colour map, PNG export, pass-prediction integration

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * NOAA APT decoder for extracting weather-satellite imagery from FM-demodulated audio.
  * Streaming-friendly resampling, rectified envelope detection, and DC-invariant sync detection for channels A/B with normalized match-quality.
  * Emits scan lines as 2,080 × 8-bit pixels with per-line normalization and original input-sample indexing.
  * Supports reset, bounded internal buffering, and preserves surplus decoded lines when output capacity is insufficient.

* **Tests**
  * End-to-end and edge-case tests covering invalid parameters, numeric behavior, chunked vs one-shot equivalence, resets, and buffer limits.

### CODE REVIEW

File: crates/sdr-dsp/src/apt.rs [hunk 1]
```diff
@@ -0,0 +1,1295 @@
...
+
+    resample_scratch: Vec<f32>,
+    envelope_scratch: Vec<f32>,
+    accumulator: Vec<f32>,
+
+    // Which input-audio sample is currently at accumulator[0]. Carried
+    // forward on every drain so we can stamp each emitted line with a
+    // source-timestamp-ish index for correlation with satellite pass
+    // ephemerides or telemetry ticks.
+    accumulator_start_input_sample: u64,
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/apt.rs]: Fixed in d173f2e. Migrated the origin field from input-rate to intermediate-rate units (`accumulator_start_intermediate_sample: u64`) — every drain advances by exact integer counts, and conversion to input-rate index now happens once at stamp time inside `accumulator_to_input_index`. Verified the math is loss-free across multi-hour captures (no floor-division leak per drain).

File: crates/sdr-dsp/src/apt.rs [hunk 2]
```diff
@@ -0,0 +1,1295 @@
...
+                f64::from(INTERMEDIATE_RATE_HZ),
+            )?,
+            envelope: EnvelopeDetector::new(INTERMEDIATE_RATE_HZ)?,
+            sync_detector: SyncDetector::new(),
+            resample_scratch: Vec::new(),
+            envelope_scratch: Vec::new(),
+            accumulator: Vec::with_capacity(DECODER_BUFFER_CAP),
+            accumulator_start_input_sample: 0,
+        })
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/apt.rs:603]: Fixed in d173f2e. Added a hard `NYQUIST_FLOOR_HZ = 4800` gate at the top of `AptDecoder::new` that returns `DspError::InvalidParameter` with a clear message before the resampler / envelope detector get touched. New `apt_decoder_rejects_sub_nyquist_input_rate` test pins the boundary.

File: crates/sdr-dsp/src/apt.rs [hunk 3]
```diff
@@ -0,0 +1,1295 @@
...
+        // from producing a high-quality match we still eat a line's worth
+        // of samples every loop iteration, but belt-and-braces: bound the
+        // worst case.
+        if self.accumulator.len() > DECODER_BUFFER_CAP {
+            let drop_n = self.accumulator.len() - DECODER_BUFFER_CAP;
+            self.accumulator.drain(..drop_n);
+            self.accumulator_start_input_sample += self.ratio_intermediate_to_input(drop_n);
+        }
+
+        Ok(lines)
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/apt.rs]: Fixed in d173f2e. Adopted the crate's standard allocation-free contract:

```rust
pub fn process(&mut self, input: &[f32], output: &mut [AptLine]) -> Result<usize, DspError>
```

Made `AptLine` slot-reusable: `pixels` changed from `Box<[u8; 2080]>` to inline `[u8; 2080]`, added `Default` so callers do `vec![AptLine::default(); N]` once and reuse forever. Decimation moved to `decimate_into_pixels` writing directly into the output slot. Added a `BufferTooSmall` error path with a regression test (`apt_decoder_buffer_too_small_returns_error`) so the contract is enforced.

File: crates/sdr-dsp/src/apt.rs [hunk 4]
```diff
@@ -0,0 +1,1295 @@
...
+        let mut lines = Vec::new();
+        while self.accumulator.len() >= MIN_ACCUMULATOR_FOR_DECODE {
+            // Search the first SAMPLES_PER_LINE tau positions — any match
+            // in there leaves a full SAMPLES_PER_LINE window for the line
+            // body without running past the end.
+            let search_len = SAMPLES_PER_LINE + SYNC_A_TEMPLATE_LEN;
+            let m = self
+                .sync_detector
+                .find_best(&self.accumulator[..search_len], SyncChannel::A)
+                .expect("search slice always longer than sync template");
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/apt.rs]: Fixed in d173f2e. Replaced `expect` with `let Some(...) else { break; }` so a future constants change (or any other breakage of the search-window invariant) gracefully exits the emit loop instead of aborting the DSP path.

File: crates/sdr-dsp/src/apt.rs [hunk 5]
```diff
@@ -0,0 +1,1295 @@
...
+    clippy::unwrap_used,
+    clippy::cast_precision_loss,
+    clippy::cast_possible_truncation,
+    clippy::cast_sign_loss,
+    clippy::float_cmp
+)]
+mod tests {
+    use super::*;
+    use core::f32::consts::TAU;
+
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/apt.rs:859]: Fixed in d173f2e. Hoisted eight `TEST_*` constants at the top of `mod tests` (input rate, realistic / odd-prime chunk sizes, output capacity, grey + gradient levels, sync quality thresholds, noise duration), each with a short rationale doc-comment. Also extracted a deduped `lcg_step` helper for the deterministic noise tests, and replaced inline magic numbers throughout the test functions.

File: crates/sdr-dsp/src/apt.rs [hunk 6]
```diff
@@ -0,0 +1,1418 @@
...
+    /// to cleanly place the 4800 Hz harmonic below Nyquist, or if the
+    /// underlying FIR / tap generation rejects the design parameters.
+    pub fn new(sample_rate_hz: u32) -> Result<Self, DspError> {
+        if f64::from(sample_rate_hz) < 2.0 * SUBCARRIER_HZ + ENVELOPE_LPF_TRANSITION_HZ {
+            return Err(DspError::InvalidParameter(format!(
+                "sample_rate_hz ({sample_rate_hz}) too low for APT envelope detection — \
+                 must be well above 2·{SUBCARRIER_HZ} Hz = {} Hz",
+                2.0 * SUBCARRIER_HZ
+            )));
+        }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/apt.rs:202]: Fixed in 39b34a7. You're right — the rectified harmonic sits at 2·f_c = 4800 Hz, so the input rate has to clear *its* Nyquist at 2·4800 = 9600 Hz. Old check was `< 5800 Hz` (just clearing the subcarrier itself), which let 8 kHz audio through silently. New gate is `<= 4·SUBCARRIER_HZ = 9600 Hz` with the reasoning spelled out in the docstring. Doesn't affect `AptDecoder` (its internal envelope detector always runs at 20800 Hz), but matters for direct callers. New `envelope_detector_rejects_below_rectified_nyquist` test pins the floor: 8 kHz fails, 9.6 kHz fails (boundary), 16 kHz passes.

File: crates/sdr-dsp/src/apt.rs [hunk 7]
```diff
@@ -0,0 +1,1418 @@
...
+            decimate_into_pixels(&self.accumulator[line_start..line_end], &mut slot.pixels);
+            slot.sync_quality = m.quality;
+            slot.sync_channel = SyncChannel::A;
+            slot.input_sample_index = self.accumulator_to_input_index(line_start);
+            produced += 1;
+
+            // Drain the accumulator through the end of the emitted line.
+            self.accumulator.drain(..line_end);
+            self.accumulator_start_intermediate_sample += line_end as u64;
+        }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/apt.rs]: Fixed in 39b34a7. Switched to streaming semantics rather than the preflight option — `process` now emits as many lines as `output` can hold and returns `Ok(produced)`, leaving any remainder buffered for the next call. State only advances for lines that actually landed in the caller's output, so retries are safe with no loss. The previous `apt_decoder_buffer_too_small_returns_error` test became `apt_decoder_undersized_output_buffers_remainder` and now does the round-trip: pass a 1-slot output, get `Ok(1)`, then call again with empty input and a roomy output, verify the buffered line genuinely surfaces. Docstring updated to spell out the streaming contract.

File: crates/sdr-dsp/src/apt.rs [hunk 8]
```diff
@@ -0,0 +1,1460 @@
...
+        // 2. Envelope-detect in place into an equally-sized scratch buffer.
+        self.envelope_scratch.resize(resampled, 0.0);
+        self.envelope.process(
+            &self.resample_scratch[..resampled],
+            &mut self.envelope_scratch,
+        )?;
+
+        // 3. Append to the accumulator.
+        self.accumulator
+            .extend_from_slice(&self.envelope_scratch[..resampled]);
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/apt.rs]: Fixed in 6b58d27. Two-part change to actually bound the design:

**1. Ready-line queue.** New `ready_lines: VecDeque<AptLine>` (capped at `READY_QUEUE_CAP = 8`, ~4 s of latency at 2 lines/sec) holds decoded-but-undelivered lines. The previous design lost those when the post-loop sample cap kicked in — they're now first-class storage that the next `process` call drains before doing anything else. Decode routing picks `output[produced]` if there's a slot, else `ready_lines.push_back` if the queue has room, else stops decoding (leaving the accumulator alone for the next call).

**2. Chunked envelope ingestion.** `process` no longer does one big `extend_from_slice` of the envelope buffer. It loops in chunks bounded by `DECODER_BUFFER_CAP`, running the decode + cap cycle between each. So even a multi-megabyte input chunk can never let `accumulator` grow past `DECODER_BUFFER_CAP + SAMPLES_PER_LINE` at any instant.

Sample-level dropping now happens **only** when *both* the live `output` and the ready queue are full — i.e. when the caller has genuinely stopped draining for multiple seconds.

Two new tests pin the contract:

* `apt_decoder_undersized_output_preserves_all_decoded_lines` (renamed from `..._buffers_remainder`) — pumps the same audio through (a) a generous-output decoder and (b) a one-slot-output decoder drained line-by-line across many `process(&[], …)` calls; asserts identical total line counts, proving no decoded data is silently dropped on the queue path.

* `apt_decoder_huge_chunk_keeps_accumulator_bounded` — pushes 100 lines through a one-slot output and asserts both `accumulator` and `ready_lines` stay at-or-below their caps after the call.

File: crates/sdr-dsp/src/apt.rs [hunk 9]
```diff
@@ -0,0 +1,1583 @@
...
+    /// detector, or tap designer.
+    pub fn new(input_rate_hz: u32) -> Result<Self, DspError> {
+        // 2 · 2400 = 4800 Hz exactly — no rounding, just hard-code so the
+        // const-context-friendly comparison below stays trivially correct.
+        const NYQUIST_FLOOR_HZ: u32 = 4_800;
+        if input_rate_hz < NYQUIST_FLOOR_HZ {
+            return Err(DspError::InvalidParameter(format!(
+                "input_rate_hz ({input_rate_hz}) must be ≥ 2·SUBCARRIER_HZ \
+                 ({NYQUIST_FLOOR_HZ}) to avoid aliasing the 2400 Hz APT subcarrier",
+            )));
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/apt.rs:571]: Fixed in 066078e. You're right — at exactly Nyquist the cosine samples hit phase-ambiguous points and collapse. Changed the guard from `<` to `<=`, updated the comment / error-message / doc `# Errors` section to call out the strict-inequality requirement, and added `assert!(AptDecoder::new(4_800).is_err())` to `apt_decoder_rejects_sub_nyquist_input_rate` so the boundary is pinned.

File: crates/sdr-dsp/src/apt.rs [hunk 10]
```diff
@@ -0,0 +1,1583 @@
...
+            resampler: RealResampler::new(
+                f64::from(input_rate_hz),
+                f64::from(INTERMEDIATE_RATE_HZ),
+            )?,
+            envelope: EnvelopeDetector::new(INTERMEDIATE_RATE_HZ)?,
+            sync_detector: SyncDetector::new(),
+            resample_scratch: Vec::new(),
+            envelope_scratch: Vec::new(),
+            accumulator: Vec::with_capacity(DECODER_BUFFER_CAP),
+            ready_lines: VecDeque::with_capacity(READY_QUEUE_CAP),
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/apt.rs]: Fixed in 066078e. Yep — round 3's chunked-ingestion design intentionally peaks the accumulator at `DECODER_BUFFER_CAP + SAMPLES_PER_LINE`, and the cap-sized initial reservation was forcing a realloc on first backpressure (and Vec keeps the larger capacity afterward — exactly the kind of stealth growth we don't want). Reserved `DECODER_BUFFER_CAP + SAMPLES_PER_LINE` up-front and added a comment explaining why. New `apt_decoder_accumulator_capacity_absorbs_intentional_overshoot` test snapshots `accumulator.capacity()`, drives backpressure with 8 lines through a 1-slot output, and asserts capacity is unchanged afterward.

File: crates/sdr-dsp/src/apt.rs [hunk 11]
```diff
@@ -0,0 +1,1627 @@
...
+            + 4;
+        self.resample_scratch.resize(est_out, 0.0);
+        let resampled = self.resampler.process(input, &mut self.resample_scratch)?;
+
+        // 2. Envelope-detect in place into an equally-sized scratch buffer.
+        self.envelope_scratch.resize(resampled, 0.0);
+        self.envelope.process(
+            &self.resample_scratch[..resampled],
+            &mut self.envelope_scratch,
+        )?;
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/apt.rs]: Fixed in c94c22b. Right call — the round 3 bounded-accumulator design was leaving the upstream stages unbounded.

Solution mirrors your suggested direction:

* New `INPUT_SUBCHUNK_SAMPLES = 8192` constant.
* New `process_subchunk` helper holds the existing resample → envelope → bounded-ingest logic.
* `process` now does `for in_chunk in input.chunks(INPUT_SUBCHUNK_SAMPLES) { self.process_subchunk(...) }`. At 48 kHz that's ~170 ms per pass / ~3550 envelope samples — well under one APT line.

Plus `new` now pre-reserves both scratch vectors for the worst-case per-subchunk output `(INPUT_SUBCHUNK_SAMPLES · INTERMEDIATE_RATE_HZ / input_rate_hz) + 4`, so the `Vec::resize` calls inside the hot path are bookkeeping-only — no realloc traffic at all in steady state.

New `apt_decoder_huge_chunk_keeps_resample_scratch_bounded` test pushes 100 lines (~9.6 MB) of audio through and asserts `resample_scratch.capacity()` and `envelope_scratch.capacity()` are both unchanged afterward, locking the bounded-memory invariant in.

File: crates/sdr-dsp/src/apt.rs [hunk 12]
```diff
@@ -0,0 +1,1720 @@
...
+    accumulator_start_intermediate_sample: u64,
+}
+
+impl AptDecoder {
+    /// Build a decoder for audio sampled at `input_rate_hz`.
+    ///
+    /// Typical value is 48000 (the output rate of the FM demodulator).
+    /// Must be at least `2 · SUBCARRIER_HZ` (4800 Hz) — below that the
+    /// 2400 Hz APT subcarrier is already aliased before this pipeline
+    /// gets a chance to look at it, which would produce silent garbage.
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-dsp/src/apt.rs]: Fixed in a7de2d8. Good catch — round 4 tightened the guard from `<` to `<=` but I'd left the prose saying "at least". Updated to **"strictly greater than"** with a one-line explanation of why 4800 Hz exactly is bad (subcarrier on Nyquist = phase-ambiguous samples on the cosine), so the docstring now matches the `# Errors` section directly below it and the actual runtime contract.
